### PR TITLE
VR Proxy implementation

### PR DESCRIPTION
**Pull request by KKarlash**
@DKlimenko 
@bonewell 
@VladSemenyuk 
@giangtqh
@thinhts

### CODE REVIEW

File: src/components/vr_cooperation/include/vr_cooperation/vr_module.h [hunk 1]
```diff
@@ -115,10 +118,10 @@ class VRModule : public functional_modules::GenericModule {
   void ReceiveMessageFromMobile();
 
   /**
-   * @brief Sends response to HMI
-   * @param msg response mesage
+   * Handles received message from HMI (Applink)
+   * @param message is GPB message according with protocol
    */
-  void SendResponseToHMI(application_manager::MessagePtr msg);
+   virtual void OnReceived(const vr_hmi_api::Message& message);
```
Comment (giangtqh) [inline src/components/vr_cooperation/include/vr_cooperation/vr_module.h]: vr_hmi_api::ServiceMessage

Comment (giangtqh) [inline src/components/vr_cooperation/include/vr_cooperation/vr_module.h]: indent not correct, remove 1 space at line start

File: src/components/vr_cooperation/include/vr_cooperation/vr_proxy.h [hunk 2]
```diff
@@ -51,16 +59,24 @@ class VRProxy {
...
-  void OnReceived();
+  void OnReceived(const vr_hmi_api::Message& message);
 
   /**
    * @brief Sends GPB message
+   * @param message to send
+   * @return true if success
    */
-  bool SendMessage();
+   bool Send(const vr_hmi_api::Message& message);
```
Comment (giangtqh) [inline src/components/vr_cooperation/include/vr_cooperation/vr_proxy.h]: remove 1 space at line start

File: src/components/vr_cooperation/include/vr_cooperation/vr_proxy.h [hunk 3]
```diff
@@ -51,16 +59,24 @@ class VRProxy {
...
 
  private:
-  DISALLOW_COPY_AND_ASSIGN(VRProxy);
+   void Handle(vr_hmi_api::Message message);
```
Comment (giangtqh) [inline src/components/vr_cooperation/include/vr_cooperation/vr_proxy.h]: remove 1 space at line start, same for line 75, 77, 79

File: src/components/vr_cooperation/include/vr_cooperation/vr_proxy_listener.h [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+#define SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_VR_PROXY_LISTENER_H_
+
+namespace vr_hmi_api {
+class Message;
+}  // namespace vr_hmi_api
+
+namespace vr_cooperation {
+
+class VRProxyListener {
+  public:
```
Comment (giangtqh) [inline src/components/vr_cooperation/include/vr_cooperation/vr_proxy_listener.h]: public: should be indented +1 space inside class

File: src/components/vr_cooperation/include/vr_cooperation/vr_proxy_listener.h [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  /**
+   * Handles receiving GPB message from HMI
+   * @param message GPB message according with hmi.proto file
+   */
+  virtual void OnReceived(const vr_hmi_api::Message& message) = 0;
+};
+
+}  // namespace vr_cooperation
+
+#endif // SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_VR_PROXY_LISTENER_H_
```
Comment (giangtqh) [inline src/components/vr_cooperation/include/vr_cooperation/vr_proxy_listener.h]: At least two spaces is best between code and comments

**Comment by giangtqh** [comment]
reviewed,
Make sure you check with CPPLint

File: src/components/vr_cooperation/include/vr_cooperation/vr_proxy.h [hunk 6]
```diff
@@ -33,16 +33,24 @@
...
 #define SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_VR_PROXY_H_
 
-#include "utils/macro.h"
+#include <queue>
+
+#include "utils/lock.h"
+#include "utils/threads/message_loop_thread.h"
+#include "vr_cooperation/interface/hmi.pb.h"
+
+#include "vr_cooperation/vr_proxy_listener.h"
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/vr_proxy.h]: Use forward declaration.

File: src/components/vr_cooperation/include/vr_cooperation/vr_proxy.h [hunk 7]
```diff
@@ -33,16 +33,24 @@
 #ifndef SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_VR_PROXY_H_
 #define SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_VR_PROXY_H_
 
-#include "utils/macro.h"
+#include <queue>
+
+#include "utils/lock.h"
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/vr_proxy.h]: You don't use lock primitives.

File: src/components/vr_cooperation/src/vr_proxy.cc [hunk 8]
```diff
@@ -37,19 +37,28 @@ namespace vr_cooperation {
...
+  listener_->OnReceived(message);
+}
+
+bool VRProxy::Send(const vr_hmi_api::Message& message) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  // TODO(KKarlash): should be implemented
+  return false;
 }
 
 }  //  namespace vr_cooperation
```
Comment (bonewell) [inline src/components/vr_cooperation/src/vr_proxy.cc]: 1 space after `//`

File: src/components/vr_cooperation/src/vr_proxy.cc [hunk 9]
```diff
@@ -37,19 +37,28 @@ namespace vr_cooperation {
 
 CREATE_LOGGERPTR_GLOBAL(logger_, "VRProxy")
 
-VRProxy::VRProxy() {
+VRProxy::VRProxy(VRProxyListener* listener)
+    : listener_(listener),
+      incoming_("VrIncoming", this){
```
Comment (bonewell) [inline src/components/vr_cooperation/src/vr_proxy.cc]: 1 space between `)` and `{`

File: src/components/vr_cooperation/interface/hmi.proto [hunk 10]
```diff
@@ -1,2 +1,5 @@
 syntax = "proto2";
-package vr_hmi_api;
\ No newline at end of file
+package vr_hmi_api;
+
+message Message {
+}
```
Comment (DKlimenko) [inline src/components/vr_cooperation/interface/hmi.proto]: please add one more empty line at the end

**Comment by DKlimenko** [comment]
reviewed

**Comment by bonewell** [comment]
Reviewed.

**Comment by VladSemenyuk** [comment]
Reviewed

File: src/components/vr_cooperation/include/vr_cooperation/vr_proxy_listener.h [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_VR_PROXY_LISTENER_H_
+#define SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_VR_PROXY_LISTENER_H_
+
+namespace vr_hmi_api {
+class ServiceMessage;
+}  // namespace hmi_api
```
Comment (giangtqh) [inline src/components/vr_cooperation/include/vr_cooperation/vr_proxy_listener.h]: // namespace vr_hmi_api

**Pull request merged**